pkg/gohai/utils/common_test.go

```diff
@@ -0,0 +1,259 @@
+// This file is licensed under the MIT License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2014-present Datadog, Inc.
+
+package utils
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// TestAsJSONTypes tests that each supported type can appear properly in the output
+func TestAsJSONTypes(t *testing.T) {
+	info := &struct {
+		SomeInt     Value[int]     `json:"my_int"`
+		SomeInt8    Value[int8]    `json:"my_int8"`
+		SomeInt16   Value[int16]   `json:"my_int16"`
+		SomeInt32   Value[int32]   `json:"my_int32"`
+		SomeInt64   Value[int64]   `json:"my_int64"`
+		SomeUint    Value[uint]    `json:"my_uint"`
+		SomeUint8   Value[uint8]   `json:"my_uint8"`
+		SomeUint16  Value[uint16]  `json:"my_uint16"`
+		SomeUint32  Value[uint32]  `json:"my_uint32"`
+		SomeUint64  Value[uint64]  `json:"my_uint64"`
+		SomeFloat32 Value[float32] `json:"my_float32"`
+		SomeFloat64 Value[float64] `json:"my_float64"`
+		SomeString  Value[string]  `json:"my_string"`
+	}{
+		SomeInt:     NewValue(1),
+		SomeInt8:    NewValue[int8](2),
+		SomeInt16:   NewValue[int16](3),
+		SomeInt32:   NewValue[int32](4),
+		SomeInt64:   NewValue[int64](5),
+		SomeUint:    NewValue[uint](6),
+		SomeUint8:   NewValue[uint8](7),
+		SomeUint16:  NewValue[uint16](8),
+		SomeUint32:  NewValue[uint32](9),
+		SomeUint64:  NewValue[uint64](10),
+		SomeFloat32: NewValue[float32](32.),
+		SomeFloat64: NewValue(64.),
+		SomeString:  NewValue("mystr"),
+	}
+
+	marshallable, warns, err := AsJSON(info, false)
+	require.NoError(t, err)
+	require.Empty(t, warns)
+
+	marshalled, err := json.Marshal(marshallable)
+	require.NoError(t, err)
+
+	// use format specifier for floats to make sure the formatting is the same
+	expected := fmt.Sprintf(`{
+		"my_int": "1",
+		"my_int8": "2",
+		"my_int16": "3",
+		"my_int32": "4",
+		"my_int64": "5",
+		"my_uint": "6",
+		"my_uint8": "7",
+		"my_uint16": "8",
+		"my_uint32": "9",
+		"my_uint64": "10",
+		"my_float32": "%f",
+		"my_float64": "%f",
+		"my_string": "mystr"
+	}`, float32(32.), 64.)
+	require.JSONEq(t, expected, string(marshalled))
+}
+
+func TestAsJSONEmpty(t *testing.T) {
+	info := &struct{}{}
+	_, _, err := AsJSON(info, false)
+	require.ErrorIs(t, err, ErrNoFieldCollected)
+
+	_, _, err = AsJSON(info, true)
+	require.ErrorIs(t, err, ErrNoFieldCollected)
+}
+
+// TestAsJsonInvalidParam tests that using something other than struct returns an error
+func TestAsJsonInvalidParam(t *testing.T) {
+	var err error
+	// function
+	f := func() {}
+	_, _, err = AsJSON(&f, false)
+	require.ErrorIs(t, err, ErrArgNotStruct)
+	// slice
+	_, _, err = AsJSON(&[]int{1}, false)
```

Comment:
Note: I don't think you need to do black box testing and just testing for one non struct type should be enough.

Reply:
I removed the weirder cases, but as some package return slices I think it's interesting to keep that case, and also the "pointer to pointer to struct" which could easily happen.